### PR TITLE
[Merged by Bors] - chore(Algebra/BigOperators/List): Use Std lemmas

### PR DESCRIPTION
**Pull request by YaelDillies**
* Make `Algebra.BigOperators.List.Basic`, `Data.List.Chain` not depend on `Data.Nat.Order.Basic` by using `Nat`-specific Std lemmas rather than general mathlib ones. I leave the `Data.Nat.Basic` import since `Algebra.BigOperators.List.Basic` is algebra territory.
* Make `Algebra.BigOperators.List.Basic` not depend on `Algebra.Divisibility.Basic`. I'm not too sure about that one since they already are algebra. My motivation is that they involve ring-like objects while big operators are about group-like objects, but this is in some sense a second order refactor.
* As a consequence, move the divisibility and `MonoidWithZero` lemmas from `Algebra.BigOperators.List.Basic` to `Algebra.BigOperators.List.Lemmas`.
* Move the content of `Algebra.BigOperators.List.Defs` to `Algebra.BigOperators.List.Basic` since no file imported the former without the latter and their imports are becoming very close after this PR.
* Make `Data.List.Count`, `Data.List.Dedup`, `Data.List.ProdSigma`, `Data.List.Zip` not depend on `Algebra.BigOperators.List.Basic`.
* As a consequence, move the big operators lemmas that were in there to `Algebra.BigOperators.List.Basic`. For the lemmas that were `Nat` -specific, keep a version of them stated using `Nat.sum`.
* To help with this, add `Nat.sum_eq_listSum (l : List Nat) : Nat.sum l = l.sum`.

---

- [x] depends on: #11741
- [x] depends on: #11750

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
I'm not convinced that introducing `Nat.sum` into lemma statements is preferable; this creates a real inconvenience (unhelpfully stated lemmas) to resolve a mostly superficial one ("too many imports"(

**Comment by YaelDillies** [comment]
Note that the lemma statements haven't changed. It's just that there are a few extra lemmas stated using `Nat.sum`

File: Mathlib/Algebra/BigOperators/List/Basic.lean [hunk 1]
```diff
@@ -619,11 +615,15 @@ theorem alternatingProd_cons (a : α) (l : List α) :
 end Alternating
 
 lemma sum_nat_mod (l : List ℕ) (n : ℕ) : l.sum % n = (l.map (· % n)).sum % n := by
-  induction l <;> simp [Nat.add_mod, *]
+  induction' l with a l ih
+  · simp only [Nat.zero_mod, map_nil]
+  · simpa only [map_cons, sum_cons, Nat.mod_add_mod, Nat.add_mod_mod] using congr((a + $ih) % n)
```
Comment (YaelDillies) [inline Mathlib/Algebra/BigOperators/List/Basic.lean:620]: `Nat.add_mod` rewrites infinitely on its own. I'm not sure how this proof ever worked.

File: Mathlib/Algebra/BigOperators/List/Basic.lean [hunk 2]
```diff
@@ -619,11 +615,15 @@ theorem alternatingProd_cons (a : α) (l : List α) :
...
 #align list.sum_nat_mod List.sum_nat_mod
 
 lemma prod_nat_mod (l : List ℕ) (n : ℕ) : l.prod % n = (l.map (· % n)).prod % n := by
-  induction l <;> simp [Nat.mul_mod, *]
+  induction' l with a l ih
+  · simp only [Nat.zero_mod, map_nil]
+  · simpa only [prod_cons, map_cons, Nat.mod_mul_mod, Nat.mul_mod_mod] using congr((a * $ih) % n)
```
Comment (YaelDillies) [inline Mathlib/Algebra/BigOperators/List/Basic.lean:626]: Similarly, simping with `Nat.mul_mod` is a terrible idea.

**Review by Ruben-VandeVelde — APPROVED** [review]
maintainer merge

**Pull request closed without merging**